### PR TITLE
react-intl upgrade

### LINKED ISSUE #1290: Update components to work with react-intl v2-v5
# Feature Request
In an endeavor to minimize the number of major version bumps we want to make the updates for our components to work with react-intl v5 passive wherever possible.

## Description
For this update the peer dependencies need to be updated. `intlShape` Is no longer exported since v3, so the imports for that must be removed and PropType for intl object updated. An undocumented non-passive change was made in react-intl for the `FormattedMessage` component, when a function is passed as a child of the component the function is now given the param of an array of strings instead of the joined string. This causes many of the components that we are rendering in these functions to have prop type warnings as they are expecting a string. There are a few approaches that we could take to make these components continue to work as expected for all the react-intl version. I believe for most of these cases we can switch from using the `FormattedMessage` component to `injectIntl` + `intl.formatMessage()` or the function provided to `FormattedMessage` can do a check if the param is an array and join the strings, if so. 

## Additional Context / Screenshots
I have begun this work at https://github.com/cerner/terra-framework/tree/react-intl-upgrade
This work should be able to be completed and released prior to the framework portions of the upgrade being completed.

In order to test these changes against react-intl v5 the package.json can be updated to pull in:
`    "terra-aggregate-translations": "git+https://github.com/cerner/terra-aggregate-translations.git#upgrade-intl",
    "terra-dev-site": "git+https://github.com/cerner/terra-dev-site.git#react-intl-upgrade",
`

### PR DESCRIPTION
**Pull request by jmsv6d**
### Summary
Made updates to components so they will be compatible with react-intl v2-v5.  

Closes #1290 

### Deployment Link

https://terra-framework-deployed-pr-#.herokuapp.com/

### Testing

Ran locally with react-intl v2 and v5, and tests pass

### Additional Details

Updates required:

- Peer dependency updated to range
- Remove `intlShape` imports as react-intl stopped exporting it in V3
- `FormattedMessage` no longer passes a string as the param for the child function, but instead an array of strings. Changes were made to either use `injectIntl` + `intl.formatMessage()` or to check if the param is an array and join the strings, to prevent propType warning for the components rendered.
- `injectIntl` wraps the component with a function component in newer versions, this caused some issues with `refs`, particularly `DatePicker` required an update to use a refCallback internally. Note: The api for `injectIntl` to forwardRef is different between v2 and v3 

Thank you for contributing to Terra.
@cerner/terra

### CODE REVIEW

File: packages/terra-abstract-modal/src/_ModalContent.jsx [hunk 1]
```diff
@@ -138,15 +138,29 @@ const ModalContent = forwardRef((props, ref) => {
         ref={ref}
       >
         <FormattedMessage id="Terra.AbstractModal.BeginModalDialog">
```
Comment (emilyrohrbough) [inline packages/terra-abstract-modal/src/_ModalContent.jsx:140]: Should we take advantage of the new fallback message logic ? Where if a translation is missing, it will fallback to the value defined here?
https://formatjs.io/docs/react-intl/components#usage

It might end up with mis-matched locales...but might be better than the translation key?

File: packages/terra-application-links/src/tabs/_TabMenu.jsx [hunk 2]
```diff
@@ -3,7 +3,7 @@ import PropTypes from 'prop-types';
 import Popup from 'terra-popup';
 import { matchPath } from 'react-router-dom';
 import * as KeyCode from 'keycode-js';
-import { FormattedMessage } from 'react-intl';
+import { injectIntl } from 'react-intl';
```
Comment (emilyrohrbough) [inline packages/terra-application-links/src/tabs/_TabMenu.jsx:6]: Would it be worth making this change throughout instead of concatenating the chunks array passed down by FormattedMessage?

File: packages/terra-collapsible-menu-view/src/CollapsibleMenuView.jsx [hunk 3]
```diff
@@ -30,6 +30,12 @@ const propTypes = {
    * Bounding container for the menu, will use window if no value provided.
    */
   boundingRef: PropTypes.func,
+
```
Comment (emilyrohrbough) [inline packages/terra-collapsible-menu-view/src/CollapsibleMenuView.jsx:33]:
```suggestion
```

File: packages/terra-date-picker/CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,9 @@
 
 ## Unreleased
 
+* Changed
+  * Updated to support a peer dependency of react-intl v2-v5\
```
Comment (emilyrohrbough) [inline packages/terra-date-picker/CHANGELOG.md:6]:
```suggestion
  * Updated to support a peer dependency of react-intl v2-v5
```

File: packages/terra-date-picker/src/DatePicker.jsx [hunk 5]
```diff
@@ -480,6 +480,7 @@ class DatePicker extends React.Component {
                 useExternalFormatMask={useExternalFormatMask}
               />
             )}
+            customInputRef="inputRefCallback"
```
Comment (emilyrohrbough) [inline packages/terra-date-picker/src/DatePicker.jsx:483]: why is this needed?

File: packages/terra-date-time-picker/src/DateTimePicker.jsx [hunk 6]
```diff
@@ -40,7 +40,7 @@ const propTypes = {
   /**
    * intl object programmatically imported through injectIntl from react-intl.
    * */
-  intl: intlShape.isRequired,
+  intl: PropTypes.shape({ locale: PropTypes.string }).isRequired,
```
Comment (emilyrohrbough) [inline packages/terra-date-time-picker/src/DateTimePicker.jsx:43]: I know it won't happen now, but do you think we should utilize react-intl's timeZone config option in the intlProvider?

File: packages/terra-abstract-modal/src/_ModalContent.jsx (hunk 1 above)
Comment (mjhenkes) [reply packages/terra-abstract-modal/src/_ModalContent.jsx:140]: Is the fallback logic available for all versions we support? IMO i'm ok if we don't fall back, if we're falling back there's something wrong with their app setup and we should make it obvious.

File: packages/terra-date-picker/src/react-datepicker/calendar.jsx [hunk 7]
```diff
@@ -112,7 +112,7 @@ export default class Calendar extends React.Component {
      * @private
      * Internationalization object with translation APIs. Provided by `injectIntl`.
      */
-    intl: intlShape,
+    intl: PropTypes.shape({}),
```
Comment (mjhenkes) [inline packages/terra-date-picker/src/react-datepicker/calendar.jsx:115]: Why are these shapes not as defined as the others?

File: packages/terra-abstract-modal/src/_ModalContent.jsx (hunk 1 above)
Comment (jmsv6d) [reply packages/terra-abstract-modal/src/_ModalContent.jsx:140]: We'll probably want to update to use that once we do next major version bumps on components, but in order to continue supporting the components to be used with the older versions of react-intl we can't.

File: packages/terra-application-links/src/tabs/_TabMenu.jsx (hunk 2 above)
Comment (jmsv6d) [reply packages/terra-application-links/src/tabs/_TabMenu.jsx:6]: I tried to mostly switch to `injectIntl`, the problem is that the wrapper that `injectIntl` creates is a functional component, so when I made that change in a few of the components it caused issues if the `ref` for the component was being used in a test or example. `react-intl` v2 has a different api for forwarding refs than later versions.

File: packages/terra-date-picker/src/DatePicker.jsx (hunk 5 above)
Comment (jmsv6d) [reply packages/terra-date-picker/src/DatePicker.jsx:483]: In later versions of `react-intl` the wrapper created by `injectIntl` is a function component. The `react-date-picker` code utilized the `ref` of the date input for focus manipulation, but the DateInput component being wrapped in a function component means it doesn't have a `ref`. There was already logic in the `react-date-picker` to pass the name of an alternative prop to pass the `ref` callback to, so I added an `inputRefCallback` prop to the DateInput that would pass it down to the `ref` of the input element and this prop indicates that's what the `react-date-picker` should use.

File: packages/terra-date-picker/src/react-datepicker/calendar.jsx (hunk 7 above)
Comment (jmsv6d) [reply packages/terra-date-picker/src/react-datepicker/calendar.jsx:115]: It's because this component doesn't actually use the `intl` object, it just passes it through as a prop. I had only defined the shape with what each component used from the `intl` object in lieu of defining the whole shape everywhere . If I dig down to where it's actually used, it looks like it's being used for the `formatMessage` function,  I can go ahead and define that here.

File: packages/terra-date-time-picker/src/DateTimePicker.jsx (hunk 6 above)
Comment (jmsv6d) [reply packages/terra-date-time-picker/src/DateTimePicker.jsx:43]: I don't see why not. We could easily add that as a prop to ApplicationBase to allow that to be set.

**Review by mjhenkes — APPROVED** [review]

File: packages/terra-application-header-layout/src/ApplicationHeaderLayout.jsx [hunk 8]
```diff
@@ -20,7 +20,7 @@ const propTypes = {
   /**
    * intl object programmatically imported through injectIntl from react-intl.
    * */
-  intl: intlShape.isRequired,
+  intl: PropTypes.shape({ formatMessage: PropTypes.func }),
```
Comment (ryanthemanuel) [inline packages/terra-application-header-layout/src/ApplicationHeaderLayout.jsx:23]: Should these be required?

**Review by ryanthemanuel — APPROVED** [review]

File: packages/terra-application-header-layout/src/ApplicationHeaderLayout.jsx (hunk 8 above)
Comment (jmsv6d) [reply packages/terra-application-header-layout/src/ApplicationHeaderLayout.jsx:23]: Oops, yeah.

File: packages/terra-application-navigation/package.json [hunk 9]
```diff
@@ -28,7 +28,7 @@
   "peerDependencies": {
     "react": "^16.8.5",
     "react-dom": "^16.8.5",
-    "react-intl": "^2.4.0",
+    "react-intl": ">=2.4.0, <6.0.0",
```
Comment (benbcai) [inline packages/terra-application-navigation/package.json:31]: Should we make this `">=2.8.0, <6.0.0"` to be consistent with the other packages?

**Review by benbcai — APPROVED** [review]
